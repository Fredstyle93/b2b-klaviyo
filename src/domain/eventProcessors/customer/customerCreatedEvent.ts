import { AbstractEvent } from '../abstractEvent';
import logger from '../../../utils/log';
import { CustomerCreatedMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';
import {
    getCTCustomerAddressForKlaviyo,
    getPhoneNumber,
    mapCTAddressToKlaviyoLocation,
} from './utils/CustomerAddressUtils';

export class CustomerCreatedEvent extends AbstractEvent {
    isEventValid(): boolean {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        return (
            customerCreatedMessage.resource.typeId === 'customer' &&
            customerCreatedMessage.type === 'CustomerCreated' &&
            !!customerCreatedMessage.customer &&
            !!customerCreatedMessage.customer.email
        );
    }

    generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as CustomerCreatedMessage;
        logger.info(`processing CT ${message.type} message`);
        const address = getCTCustomerAddressForKlaviyo(message.customer);
        const body: ProfileRequest = {
            data: {
                type: 'profile',
                attributes: {
                    email: message.customer.email,
                    first_name: message.customer.firstName,
                    last_name: message.customer.lastName,
                    title: message.customer.title,
                    phone_number: getPhoneNumber(address?.phone),
                    organization: message.customer.companyName,
                    location: mapCTAddressToKlaviyoLocation(address),
                },
                meta: {
                    identifiers: {
                        external_id: message.resource.id,
                    },
                },
            },
        };
        return Promise.resolve([
            {
                body,
                type: 'profileUpdated',
            },
        ]);
    }
}
