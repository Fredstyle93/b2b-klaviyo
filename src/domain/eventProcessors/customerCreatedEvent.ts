import { AbstractEvent } from './abstractEvent';
import logger from '../../utils/log';
import { CustomerCreatedMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';
import { Address } from '@commercetools/platform-sdk';

export class CustomerCreatedEvent extends AbstractEvent {
    isEventValid(): boolean {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        return (
            customerCreatedMessage.resource.typeId === 'customer' &&
            customerCreatedMessage.type === 'CustomerCreated' &&
            !!customerCreatedMessage.customer &&
            !!customerCreatedMessage.customer.email
        );
    }

    generateKlaviyoEvents(): KlaviyoEvent[] {
        logger.info('processing CT customer created message');
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        const address = this.getCTCustomerAddress();
        const body = {
            data: {
                type: 'profile',
                attributes: {
                    external_id: customerCreatedMessage.customer.id,
                    email: customerCreatedMessage.customer.email,
                    first_name: customerCreatedMessage.customer.firstName,
                    last_name: customerCreatedMessage.customer.lastName,
                    title: customerCreatedMessage.customer.title,
                    phone_number: address?.phone,
                    organization: customerCreatedMessage.customer.companyName,
                    location: {
                        address1: this.getAddress1(address),
                        address2: this.getAddress2(address),
                        city: address?.city,
                        country: address?.country,
                        // latitude: null,
                        // longitude: null,
                        region: address?.region,
                        zip: address?.postalCode,
                        // timezone: null,
                    },
                },
            },
        };
        return [
            {
                body,
                type: 'profile',
            },
        ];
    }

    getCTCustomerAddress(): Address | undefined {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        if (customerCreatedMessage.customer.defaultBillingAddressId) {
            return customerCreatedMessage.customer.addresses.find(
                (address) => address.id === customerCreatedMessage.customer.defaultBillingAddressId,
            );
        }
        if (
            customerCreatedMessage.customer.billingAddressIds &&
            customerCreatedMessage.customer.billingAddressIds.length
        ) {
            return customerCreatedMessage.customer.addresses.find(
                (address) => address.id === customerCreatedMessage?.customer?.billingAddressIds?.find(() => true),
            );
        }
        return customerCreatedMessage.customer.addresses.find(() => true);
    }

    getAddress1(address?: Address): string | undefined {
        return address
            ? [address.apartment, address.building, address.streetNumber, address.streetName]
                  .filter((element) => element)
                  .join(', ')
            : undefined;
    }

    getAddress2(address?: Address): string | undefined {
        return address
            ? [address.additionalStreetInfo, address.additionalAddressInfo].filter((element) => element).join(', ')
            : undefined;
    }
}
