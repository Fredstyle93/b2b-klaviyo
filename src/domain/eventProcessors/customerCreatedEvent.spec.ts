import { CustomerCreatedEvent } from './customerCreatedEvent';
import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { mockDeep } from 'jest-mock-extended';
import { expect as exp } from 'chai';
import { sampleCustomerCreatedEvent } from '../../test/testData/customerData';

describe('customerCreatedEvent > isEventValid', () => {
    it('should return valid when the customer created resource has all the required fields', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'customer' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'CustomerCreated' });
        Object.defineProperty(ctMessageMock, 'customer', { value: { email: 'some@email.com' } });

        const event = CustomerCreatedEvent.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource      | type                 | customer
        ${'invalid'}  | ${'CustomerCreated'} | ${{ email: 'someemail@klaviyo.com' }}
        ${'customer'} | ${'invalid'}         | ${{ email: 'someemail@klaviyo.com' }}
        ${'customer'} | ${'CustomerCreated'} | ${null}
    `('should return invalid when is not a valid customer created message', ({ resource, type, customer }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } });
        Object.defineProperty(ctMessageMock, 'type', { value: type });
        Object.defineProperty(ctMessageMock, 'customer', { value: customer });

        const event = CustomerCreatedEvent.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('customerCreatedEvent > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event when the input customer created event is valid', () => {
        const message = sampleCustomerCreatedEvent as unknown as MessageDeliveryPayload;
        const event = CustomerCreatedEvent.instance(message);

        const klaviyoEvent = event.generateKlaviyoEvent();
        exp(klaviyoEvent).to.not.be.undefined;
        expect(klaviyoEvent.body).toMatchSnapshot();
    });
});
