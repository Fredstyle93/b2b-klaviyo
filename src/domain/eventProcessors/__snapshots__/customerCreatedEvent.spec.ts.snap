// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo event when the input customer created event is valid 1`] = `
{
  "data": {
    "attributes": {
      "email": "rob.smith@e2x.com",
      "external_id": "2925dd3a-5417-4b51-a76c-d6721472531f",
      "first_name": "Roberto",
      "last_name": "Smith",
      "location": {
        "address1": "C, Tall Tower, 23, High Road",
        "address2": "private access, additional address info",
        "city": "London",
        "country": "UK",
        "region": "aRegion",
        "zip": "WE1 2DP",
      },
      "organization": "Klaviyo",
      "phone_number": "+44 0128472834",
      "title": "Mr",
    },
    "type": "profile",
  },
}
`;
